compiler: env/path handling fixes

tamird opened the pull request:
Please see individual commits. r? @bjorn3 cf. #116426

tamird committed:
rustc_driver: avoid fallible conversions

Use `std::path::PathBuf` rather than `String`; use `std::env::var_os`
rather than `std::env::var`. These changes avoid a number of error paths
which can arise in the presence of non-UTF-8 paths.

tamird committed:
rustc_codegen_ssa: use try_canonicalize in rpath

This is simpler and avoids unnecessary calls to `env::current_dir`.

tamird committed:
rustc_metadata: use try_canonicalize

This is simpler and avoids unnecessary calls to `env::current_dir`.

rustc_plugin is left unchanged to avoid conflicts with #116412.

Updates #116426.

bjorn3 commented on compiler/rustc_driver_impl/src/lib.rs:
```diff
@@ -1287,17 +1287,21 @@ pub fn ice_path() -> &'static Option<PathBuf> {
             return None;
         }
-        if let Ok("0") = std::env::var("RUST_BACKTRACE").as_deref() {
+        if let Some(s) = std::env::var_os("RUST_BACKTRACE") && s == "0" {
```
There should be no difference here. `env::var` will return `Err` in case of a non-UTF-8 value and `env::var_os` will return a value not comparing equal to `"1"`.

tamird replied:
I agree, but this avoids the appearance of an unhandled error.

Would you prefer I revert this?

bjorn3 replied:
Yes, please revert this and elsewhere.

tamird replied:
Actually, no - this is different; the old code would treat non-UTF8 paths as invalid and call env::current_dir. This now correctly handles non-unicode paths.

bjorn3 replied:
The place where a path is passed in an env var can remain the same, but the rest can be reverted without behavior changes.

tamird replied:
You're right - I went ahead and change all interactions with `RUST_BACKTRACE` to `env_os` for consistency and updated the commit message.

bjorn3 commented on compiler/rustc_codegen_ssa/src/back/rpath.rs:
```diff
@@ -82,12 +80,11 @@ fn get_rpath_relative_to_output(config: &mut RPathConfig<'_>, lib: &Path) -> OsS
+    let lib = lib.parent().unwrap();
+    let output = config.out_filename.parent().unwrap();
+    let lib = lib.canonicalize().unwrap();
+    let output = output.canonicalize().unwrap();
```
Canonicalization can fail and on Windows there is a ramdisk implementation where it always fails.

ChrisDenton replied:
We now have `try_canonicalize` in [`rustc_fs_util`](https://github.com/rust-lang/rust/blob/d4ba2b4c7c938cf90c03a265cb31356537f608ad/compiler/rustc_fs_util/src/lib.rs#L97). It tries to get the canonical path and falls back to merely getting an absolute path.

tamird replied:
Nice! Done.

tamird committed:
compiler: always use var_os("RUST_BACKTRACE")

There are 3 instances of var(...) and 3 instances of var_os(...); the
latter avoids an appearance of unhandled error, so use it everywhere.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6431756044/job/17465271195) [(plain)](https://github.com/rust-lang/rust/commit/238b52ba6afad10f55c1f3e5afc9807e80e4ab75/checks/17465271195/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=tamird
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_ef8ebaf3-2ed6-4ec9-99b2-28a862e8627c
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=avoid-unwrap-absolute
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_ef8ebaf3-2ed6-4ec9-99b2-28a862e8627c
GITHUB_REF=refs/pull/116487/merge
GITHUB_REF_NAME=116487/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=ee9cbf0920722cbbb4d7f6e9a6546468bfdcf638
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_ef8ebaf3-2ed6-4ec9-99b2-28a862e8627c
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_ef8ebaf3-2ed6-4ec9-99b2-28a862e8627c
GITHUB_TRIGGERING_ACTOR=tamird
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116487/merge
GITHUB_WORKFLOW_SHA=ee9cbf0920722cbbb4d7f6e9a6546468bfdcf638
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---

failures:

---- back::rpath::tests::test_rpath_relative stdout ----
thread 'back::rpath::tests::test_rpath_relative' panicked at compiler/rustc_codegen_ssa/src/back/rpath.rs:86:34:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }


error: test failed, to rerun pass `-p rustc_codegen_ssa --lib`
failures:
    back::rpath::tests::test_rpath_relative

test result: FAILED. 115 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.88ms
```

</details>

tamird commented on compiler/rustc_plugin_impl/src/load.rs:
```diff
@@ -60,11 +60,19 @@ fn load_plugin(
+    };
 
-    let lib = unsafe { Library::new(&lib_path) }?;
+    let lib = unsafe { Library::new(lib_path) }?;
```
Reverted and updated the commit message to indicate why.

bjorn3 commented on compiler/rustc_codegen_cranelift/build_system/main.rs:
```diff
@@ -134,8 +134,7 @@ fn main() {
 
-    let current_dir = std::env::current_dir().unwrap();
-    out_dir = current_dir.join(out_dir);
+    out_dir = rustc_fs_util::try_canonicalize(out_dir).unwrap();
```
This won't work here. Rustc_codegen_cranelift's build systen doesn't and can't depend on any rustc crate.

tamird replied:
Thanks, and indeed `current_dir` is used below. Removed this commit.

tamird commented:
@bjorn3 ty for the reviews! Anything more you'd like changed?

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit 3cac3de200a99ed29583ce95705696e4b4782c76 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 3cac3de200a99ed29583ce95705696e4b4782c76 -->
<!-- homu: {"type":"Approved","sha":"3cac3de200a99ed29583ce95705696e4b4782c76","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 3cac3de200a99ed29583ce95705696e4b4782c76 with merge e08de860365aa73653dc5fc45b09f194785f3e57...
<!-- homu: {"type":"BuildStarted","head_sha":"3cac3de200a99ed29583ce95705696e4b4782c76","merge_sha":"e08de860365aa73653dc5fc45b09f194785f3e57"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6445749311/job/17501075014)
Approved by: bjorn3
Pushing e08de860365aa73653dc5fc45b09f194785f3e57 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6445749311/job/17501075014"},"merge_sha":"e08de860365aa73653dc5fc45b09f194785f3e57"} -->

rust-timer commented:
Finished benchmarking commit (e08de860365aa73653dc5fc45b09f194785f3e57): [comparison URL](https://perf.rust-lang.org/compare.html?start=1516ca1bc01181af73c7b7760fb90137007da75f&end=e08de860365aa73653dc5fc45b09f194785f3e57&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=1516ca1bc01181af73c7b7760fb90137007da75f&end=e08de860365aa73653dc5fc45b09f194785f3e57&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1516ca1bc01181af73c7b7760fb90137007da75f&end=e08de860365aa73653dc5fc45b09f194785f3e57&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 5.3%  | [5.3%, 5.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.6%, -2.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-1.3%, -1.3%] | 1     |
| All ❌✅ (primary)                 | -2.6% | [-2.6%, -2.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1516ca1bc01181af73c7b7760fb90137007da75f&end=e08de860365aa73653dc5fc45b09f194785f3e57&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=1516ca1bc01181af73c7b7760fb90137007da75f&end=e08de860365aa73653dc5fc45b09f194785f3e57&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 623.926s -> 621.876s (-0.33%)
**Artifact size**: 270.72 MiB -> 270.72 MiB (-0.00%)
